micro_http/src/connection.rs

```diff
@@ -0,0 +1,634 @@
+                        }
+
+                        break;
+                    } else {
+                        // Append only the remaining necessary bytes to the body of the request.
+                        self.body_vec.extend_from_slice(
+                            &self.buffer[line_start_index
+                                ..(line_start_index + self.body_bytes_to_be_read as usize)],
+                        );
+                        line_start_index += self.body_bytes_to_be_read as usize;
+                        self.body_bytes_to_be_read = 0;
+                    }
+
+                    let request = self.request.as_mut().unwrap();
+                    // If there are no more bytes to be read for this request.
+                    if self.body_bytes_to_be_read == 0 {
+                        // Assign the body of the request.
+                        let placeholder: Vec<_> = self
+                            .body_vec
+                            .drain(..request.headers.content_length() as usize)
+                            .collect();
+                        request.body = Some(Body::new(placeholder));
+
+                        // If we read more bytes than we should have into the body of the request.
+                        if !self.body_vec.is_empty() {
+                            return Err(ConnectionError::StreamError);
+                        }
+
+                        self.state = ConnectionState::RequestReady;
+                    }
+                }
+                ConnectionState::RequestReady => {
+                    // This request is ready to be passed for handling.
+                    // Update the state machine to expect a new request and push this request into
+                    // the `parsed_requests` queue.
+                    self.state = ConnectionState::WaitingForRequestLine;
+                    self.body_bytes_to_be_read = 0;
+                    self.parsed_requests.push_back(self.request.take().unwrap());
+                }
+            };
+        }
+
+        Ok(())
+    }
+
+    /// Tries to write responses to the provided stream.
+    /// Meant to be used only with non-blocking streams and an `EPOLL` structure.
+    /// Should be called whenever an `EPOLLOUT` event is signaled.
+    pub fn try_write(&mut self) -> Result<(), ConnectionError> {
+        if self.response_buffer.is_none() {
+            if let Some(response) = self.response_queue.pop_front() {
+                let mut response_buffer_vec: Vec<u8> = Vec::new();
+                response
+                    .write_all(&mut response_buffer_vec)
+                    .map_err(|_| ConnectionError::StreamError)?;
+                self.response_buffer = Some(response_buffer_vec);
+            } else {
+                return Ok(());
+            }
+        }
+
+        let mut response_fully_written = false;
+
+        if let Some(response_buffer_vec) = self.response_buffer.as_mut() {
```

Comment:
This `if` is always true. Above block either returns or populates `response_buffer`. Suggest replacing with

```
let response_buffer_vec = self.response_buffer.as_mut().unwrap();
```

Reply:
This was done as to scope the mutable borrow and to be able to make `response_buffer` a `None` after we are done writing it.